server/datastore/mysql/vpp.go

```diff
@@ -667,79 +626,243 @@ func (ds *Datastore) InsertVPPToken(ctx context.Context, tok *fleet.VPPTokenData
+	}
+
+	var tokRaw fleet.VPPTokenRaw
+	if err := json.Unmarshal(tokRawBytes, &tokRaw); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "unmarshalling raw vpp token")
+	}
+
+	exp, err := time.Parse(fleet.VPPTimeFormat, tokRaw.ExpDate)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "parsing vpp token expiration date")
+	}
+	exp = exp.UTC()
+
+	vppTokenDB := &fleet.VPPTokenDB{
+		OrgName:   tokRaw.OrgName,
+		Location:  tok.Location,
+		RenewDate: exp,
+		Token:     tok.Token,
+	}
+
+	return vppTokenDB, nil
+}
+
+func (ds *Datastore) GetVPPToken(ctx context.Context, tokenID uint) (*fleet.VPPTokenDB, error) {
 	stmt := `
-	UPDATE
-		vpp_tokens
-	SET
-		team_id = ?,
-		null_team_type = ?
+	SELECT
+		id,
+		organization_name,
+		location,
+		renew_at,
+		token
+	FROM
+		vpp_tokens v
 	WHERE
 		id = ?
 `
-	if teamID != nil && nullTeam != fleet.NullTeamNone {
-		return ctxerr.Errorf(ctx, "NullTeam must be set to NullTeamNone if TeamID is present")
+	stmtTeams := `
+	SELECT
+		vt.team_id,
+		vt.null_team_type,
+		COALESCE(t.name, '') AS name
+	FROM
+		vpp_token_teams vt
+	LEFT OUTER JOIN
+		teams t
+	ON t.id = vt.team_id
+	WHERE
+		vpp_token_id = ?
+`
+
+	var tokEnc struct {
+		ID        uint      `db:"id"`
+		OrgName   string    `db:"organization_name"`
+		Location  string    `db:"location"`
+		RenewDate time.Time `db:"renew_at"`
+		Token     []byte    `db:"token"`
 	}
 
-	if err := ds.checkVPPNullTeam(ctx, &id, nullTeam); err != nil {
-		return ctxerr.Wrap(ctx, err, "checking vpp null team for update")
+	var tokTeams []struct {
+		TeamID   *uint              `db:"team_id"`
+		NullTeam fleet.NullTeamType `db:"null_team_type"`
+		Name     string             `db:"name"`
 	}
 
-	if _, err := ds.writer(ctx).ExecContext(ctx, stmt, teamID, nullTeam, id); err != nil {
-		return ctxerr.Wrap(ctx, err, "updating vpp token team")
+	if err := sqlx.GetContext(ctx, ds.reader(ctx), &tokEnc, stmt, tokenID); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ctxerr.Wrap(ctx, notFound("VPPToken"), "selecting vpp token from db")
+		}
+		return nil, ctxerr.Wrap(ctx, err, "selecting vpp token from db")
 	}
 
-	return nil
+	tokDec, err := decrypt(tokEnc.Token, ds.serverPrivateKey)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "decrypting vpp token with serverPrivateKey")
+	}
+
+	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &tokTeams, stmtTeams, tokenID); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "selecting vpp token teams from db")
+	}
+
+	tok := &fleet.VPPTokenDB{
+		ID:        tokEnc.ID,
+		OrgName:   tokEnc.OrgName,
+		Location:  tokEnc.Location,
+		RenewDate: tokEnc.RenewDate,
+		Token:     string(tokDec),
+	}
+
+	if tokTeams == nil {
+		// Not assigned, no need to loop over teams
+		return tok, nil
+	}
+
+TEAMLOOP:
+	for _, team := range tokTeams {
+		switch team.NullTeam {
+		case fleet.NullTeamAllTeams:
+			// This should only be possible if there are no other teams
+			// Make sure something array is non-nil
```

Comment:
I thought about that as well. We probably should, although it will stop you from being able to see the token in the front end (I think)

Reply:
If we can't return a list of tokens, how would a user be able to delete the broken token?

Reply:
Not sure what's the best way to handle that, it's something that should never happen but that technically could, and it should be raised so that it can be solved... Do we have access to a logger in the Datastore? I don't remember. That could be a simple first step to make that inconsistency discoverable.

Reply:
Fixed